content/courses/advanced-ada/parts/control_flow/exceptions.rst

```diff
@@ -3,19 +3,209 @@ Exceptions
 
 .. include:: ../../../global.txt
 
-..
-    TO BE DONE:
 
-    Classification of Errors
-    ------------------------
+.. _Adv_Ada_Classification_Of_Errors:
 
-    .. admonition:: In the Ada Reference Manual
+Classification of Errors
+------------------------
 
-        - :arm:`1.1.5 Classification of Errors <1-1-5>`
+When we talk about errors and erroneous behavior in Ada, we can classify them
+in one of the four categories:
 
-    .. todo::
+- compilation errors |mdash| i.e. errors that an Ada compiler must detect at
+  compilation time;
+
+- runtime errors |mdash| i.e. errors that are detected by an Ada-based
+  application using checks at runtime;
+
+- bounded errors;
+
+- erroneous execution.
+
+In this section, we discuss each of these categories.
+
+.. admonition:: In the Ada Reference Manual
+
+    - :arm:`1.1.5 Classification of Errors <1-1-5>`
+
+
+.. _Adv_Ada_Compilation_Errors:
+
+Compilation errors
+~~~~~~~~~~~~~~~~~~
+
+In the category of compilation errors, the goal is to prevent compilers to
+accept illegal programs. Here, any program that doesn't follow the rules
+described in the Ada Reference Manual is considered illegal. Those rules
+include not only simple syntax errors, but also more complicated rules, such as
+the ones concerning
+:ref:`accessibility levels <Adv_Ada_Accessibility_Levels_Intro>` for access
+types.
+
+Note that Ada |mdash| in contrast to many programming languages, which can be
+quite permissive |mdash| tries to prevent as many errors as possible at
+compilation time because of its focus on safety. However, even though a wide
+range of errors can be detected at compilation time, this doesn't mean that a
+legal Ada program is free of errors. Therefore, using methods such as static
+analysis or unit testing is important.
+
+
+.. _Adv_Ada_Runtime_Errors:
+
+Runtime errors
+~~~~~~~~~~~~~~
+
+When a rule cannot be verified at compilation time, a common strategy is to
+have the compiler insert runtime checks into the resulting application. We see
+details about these checks later on when we discuss
+:ref:`checks and exceptions <Adv_Ada_Checks_And_Exceptions>`.
+
+A typical example is an :ref:`overflow check <Adv_Ada_Overflow_Check>`.
+Consider a calculation using variables: if this calculation leads to a result
+that isn't representable with the underlying data types, we cannot possibly
+store a value  into a register or memory that can be considered correct |mdash|
+so we have to detect this situation. Unfortunately, because we're using
+variables, we obviously cannot verify the result of he calculation at
+compilation time, so we have to verify it at runtime.
```

Comment:
"of he calculation"  => "of the calculation"

Reply:
Thanks for noticing!